lint/registration.go

```diff
@@ -0,0 +1,350 @@
+}
+
+// lintNamesToMap converts a list of lit names into a bool hashmap useful for
+// filtering. If any of the lint names are not known by the registry an error is
+// returned.
+func (r *registryImpl) lintNamesToMap(names []string) (map[string]bool, error) {
+	if len(names) == 0 {
+		return nil, nil
+	}
+
+	namesMap := make(map[string]bool, len(names))
+	for _, n := range names {
+		n = strings.TrimSpace(n)
+		if l := r.ByName(n); l == nil {
+			return nil, fmt.Errorf("unknown lint name %q", n)
+		}
+		namesMap[n] = true
+	}
+	return namesMap, nil
+}
+
+func sourceListToMap(sources SourceList) map[LintSource]bool {
+	if len(sources) == 0 {
+		return nil
+	}
+	sourceMap := make(map[LintSource]bool, len(sources))
+	for _, s := range sources {
+		sourceMap[s] = true
+	}
+	return sourceMap
+}
+
+// Filter creates a new Registry with only the lints that meet the FilterOptions
+// criteria included.
+//
+// FilterOptions are applied in the following order of precedence:
+//   ExcludeSources > IncludeSources > NameFilter > ExcludeNames > IncludeNames
+func (r *registryImpl) Filter(opts FilterOptions) (Registry, error) {
+	// If there's no filtering to be done, return the existing Registry.
+	if opts.Empty() {
+		return r, nil
+	}
+
+	filteredRegistry := newRegistry()
+
+	sourceExcludes := sourceListToMap(opts.ExcludeSources)
+	sourceIncludes := sourceListToMap(opts.IncludeSources)
+
+	nameExcludes, err := r.lintNamesToMap(opts.ExcludeNames)
+	if err != nil {
+		return nil, err
+	}
+	nameIncludes, err := r.lintNamesToMap(opts.IncludeNames)
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.NameFilter != nil && (len(nameExcludes) != 0 || len(nameIncludes) != 0) {
+		return nil, errors.New(
+			"FilterOptions.NameFilter cannot be used at the same time as " +
+				"FilterOptions.ExcludeNames or FilterOptions.IncludeNames")
+	}
+
+	for _, name := range r.Names() {
+		l := r.ByName(name)
+
+		if sourceExcludes != nil && sourceExcludes[l.Source] {
+			continue
+		}
+		if sourceIncludes != nil && !sourceIncludes[l.Source] {
+			continue
+		}
+		if opts.NameFilter != nil && !opts.NameFilter.MatchString(name) {
+			continue
+		}
+		if nameExcludes != nil && nameExcludes[name] {
+			continue
+		}
+		if nameIncludes != nil && !nameIncludes[name] {
+			continue
+		}
+
+		// when adding lints to a filtered registry we do not want Initialize() to
+		// be called a second time, so provide false as the initialize argument.
+		if err := filteredRegistry.register(l, false); err != nil {
+			return nil, err
+		}
+	}
+
+	return filteredRegistry, nil
+}
+
+// WriteJSON writes a description of each registered lint as
+// a JSON object, one object per line, to the provided writer.
+func (r *registryImpl) WriteJSON(w io.Writer) {
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(false)
+	for _, name := range r.Names() {
+		_ = enc.Encode(r.ByName(name))
+	}
+}
+
+// newRegistry constructs a registryImpl that can be used to register lints.
+func newRegistry() *registryImpl {
```

Comment:
I would make this function public, to make it easier for third-parties to create a registry of custom lints. AFAICT, right now to do so, they would have to reimplement Registry fully, or modify the global registry.

I like having the global registry since it's what most people want, most of the time, but having a separate registry also seems useful.